Upgrade yeoman-generator to version 0.19.1

alfonso-presa opened the pull request:
I've tried to upgrade the generator to the 0.19.1 API as it was using fs directly.

Let me know if there's something wrong.

BTW, I tried to use sinom to spy on the generator.log.error method for testing asserts but it seems to be already wrapped.

tomwayson commented:
@alfonso-presa we should probably also replace calls to `this.template()` w/ `this.fs.copyTmpl()` right?

tomwayson commented:
@alfonso-presa 

I made a PR to your UpgradeGeneratorVersion branch which includes a couple of extra changes that I *think* help bring this more in line w/ the way things are supposed to work in yeoman 0.19, but I'll let @SBoudrias weigh in on whether or not I've got it right:

https://github.com/alfonso-presa/generator-karma/pull/1

alfonso-presa commented:
Included @tomwayson changes and @SBoudrias comments to his PR. Thank you guys.

kevva commented on app/index.js:
```diff
@@ -166,144 +165,120 @@ module.exports = yeoman.generators.Base.extend({
-        this.log.error('Could not parse package.json.', err);
-        done();
+      var data = this.fs.readJSON(this.destinationPath('package.json'));
+      if(!data) {
```
`if (!data)`

SBoudrias commented on app/index.js:
```diff
@@ -166,144 +165,120 @@ module.exports = yeoman.generators.Base.extend({
+        this.destinationPath(path.join(this.options['config-path'], this.options['config-file'])),
+        {
+          pkg: this.pkg,
+          basePath: this.options["base-path"],
```
Use single quote.

SBoudrias commented on app/index.js:
```diff
@@ -166,144 +165,120 @@ module.exports = yeoman.generators.Base.extend({
-        data.devDependencies[plugin] = '*';
-      })
-      data.devDependencies = sortedObject(data.devDependencies);
+        this.options.plugins.forEach(function (plugin) {
```
This indentation looks wrong, seems to be one off.

SBoudrias commented on app/index.js:
```diff
@@ -166,144 +165,120 @@ module.exports = yeoman.generators.Base.extend({
-
-      fs.writeFile(packageJson, JSON.stringify(data, null, 2), done);
-    }.bind(this));
+      else{
```
The correct structure of if/else block according to our styleguide is:

```js
if () {

} else {

}
```

SBoudrias commented on app/index.js:
```diff
@@ -166,144 +165,120 @@ module.exports = yeoman.generators.Base.extend({
-        this.npmInstall();
-    }
+  install: function () {
+    this.npmInstall(undefined,undefined,this.async());
```
Don't pass any argument here.

Also note that passing `this.async()` callback here will deadlock the node process and just close the generator. - So this is actually breaking the generator :P

SBoudrias commented on test/test-creation.js:
```diff
@@ -50,24 +50,18 @@ describe('Karma generator creation test', function () {
-        gen.withOptions({travis: true}).on('end', function () {
+        var logMessage = '';
+        gen.withOptions({travis: true})
+        .on('ready', function() {
```
Indentation of this block seems off. It would need one more.

SBoudrias commented on test/test-creation.js:
```diff
@@ -75,31 +69,34 @@ describe('Karma generator creation test', function () {
-        done();
-      });
+    helpers
+    .run(join(__dirname, '../app'))
```
Indentation of these blocks seems off. It would need one more.

SBoudrias commented:
Added some comments about the code style and a bug I found.

alfonso-presa commented:
@SBoudrias, thanks for the feedback. They should all be corrected in the last commit. 

About the async()... I was expecting this to be necessary for the generator not to trigger the "end" event until the instalations have ended. In the past I've seen the "npm install" running at the same time when using one generator inside another (i.e. generator-angular) causing concurrency issues....is this solved in generator-yeoman 0.19?

SBoudrias commented:
It's been solved for a couple versions now. But the deadlocking issue is only in yeoman-generator since 0.18

alfonso-presa commented:
Ok, thanks! Great to hear it's solved :-). I removed the async call in the last commit, so I guess everything is ok now.

SBoudrias commented on app/index.js:
```diff
@@ -2,13 +2,12 @@
 module.exports = yeoman.generators.Base.extend({
-  init: function () {
+  initializing: function () {
     var notEmpty = function (str) {
```
Can we rename this function to something explicit? `notEmpty` sounds like a boolean check, not an array filtering method...

Maybe `stringArray` or `arrayFromString`

addyosmani replied:
:+1: arrayFromString

SBoudrias commented on app/index.js:
```diff
@@ -166,144 +165,119 @@ module.exports = yeoman.generators.Base.extend({
+      var data = this.fs.readJSON(this.destinationPath('package.json'));
+
+      if (!data) {
+        this.log.error('Could not open package.json for reading.', {});
```
no need for the empty object `{}`

SBoudrias commented on test/test-creation.js:
```diff
@@ -50,56 +50,53 @@ describe('Karma generator creation test', function () {
+    helpers
+      .run(join(__dirname, '../app'))
+      .withOptions({travis: true, force: true})
+      .inDir(join(__dirname, 'temp'), function(dir){
```
`function () {`

SBoudrias replied:
let's use `inTmpDir()`

SBoudrias commented:
Some new comments, went through the whole thing this time so it should be the last iteration of review comments :)

addyosmani commented:
Minus the code style comments, LGTM2.

alfonso-presa commented:
Sorry for the delay... here you have the changes.

By the way.. I've seen that when using this inside other generators based in older versions of yeoman-generator (for example generator-angular) it now displays conflict when generator-karma is editing package.json (even if the folder was initially empty).... Is there any way to solve this different than upgrading that generator to 0.19.x?

alfonso-presa commented:
Looks like using inTmpDir breaks the build in travis due to a failing test. My guess is that in travis it's using the same tmp folder and it's not clearing it before running the tests. This is the failing build:
https://travis-ci.org/yeoman/generator-karma/builds/59006966

Changing back to inDir solved the issue...

alfonso-presa commented:
Ok... it was my fault. For some reason in travis (probably because it's slower than my machine) it was ending the inTmpDir before it appends the onready event that adds the spy to the log error in the test. I rewrote the test to ensure the onready is attached.

I have already squashed the commits in case you wan't to merge.

SBoudrias commented:
Thanks a lot!!

alfonso-presa commented:
Thank you!
